Scrape Party List Members

ondenman opened the pull request:
The 2012 term differs from the 2008 and 2016 terms because it
contains party list members. The party list members are listed
in a table with a layout that differs from that of the constituency
list members.

Previously, the constituency name for each party list member was being set as that member's personal name.
The new class returns nil for this field as party list members do not have a constituency.

The class inherits directly from NokogiriDocument as fields created with FieldSerializer cannot be inherited currently.

ondenman committed:
Add PartyListKhuralMember

Class added to represent member elected from party list.

The 2012 term differs from the 2008 and 2016 terms because it
contains party list members. The party list members are listed
in a table with a layout that differs from that of the constituency
list members. Previously, the constituency name for each party list
member was being set as that member's personal name.The new class returns nil for this field as party list members do not have a constituency.

The class inherits directly from NokogiriDocument as fields
created with FieldSerializer cannot be inherited currently.

ondenman committed:
Inject member class into TermPage

TermPage is initialized with a table node. But those table nodes
differ between member types (constituency and party list members).

A member class is passed along with the table so that the correct
member type is used for that table.

tmtmtmtm commented on lib/member_table.rb:
```diff
@@ -1,17 +1,29 @@
+
+  def string_class
+    String
+  end
```
what is this method for? I can't see how it's used.

tmtmtmtm commented on lib/term_page.rb:
```diff
@@ -3,12 +3,25 @@
 class TermPage < NokogiriDocument
   field :members do
-    MemberTable.new(table).members
+    MemberTable.new(table: constituency_table, member_class: constituency).members |
```
`|` seems a little surprising here. Should this just be `+`?

tmtmtmtm commented on lib/term_page.rb:
```diff
@@ -3,12 +3,25 @@
+
+  def constituency
+    KhuralMember
+  end
```
This is quite a confusing method. What is the `constituency` of a TermPage?

tmtmtmtm commented on lib/member_table.rb:
```diff
@@ -1,17 +1,29 @@
 
 class MemberTable < NokogiriDocument
+  def initialize(table:, member_class:)
+    super(table)
```
It seems like quite a bad idea to have `new` in a subclass have an entirely different signature to the parent class. If you really need to switch to named parameters here (and I'm not sure that you do), then you should probably change NokogiriDocument to use the same approach everywhere.

ondenman committed:
Remove unneeded string_class function

Removed as it serves no purpose.

ondenman committed:
Move member_class parameter to NokogiriDocument

MemberTable needs to know about @member_class but passing it to #initialize
means the signature of the constructor is different from that of the
its parent.

ondenman committed:
Rename #constituency to #constituency_member_class

New method name is better description of what method does.

ondenman committed:
Rename #party_list to #party_list_member_class

New method name is better description of what method does.

ondenman committed:
Concatenate lists of party and constituency members

Uses #+ to concatenate the two lists rather than #| because
the two lists are entirely separate entities and we're not interested
in removing duplicates.

tmtmtmtm commented on lib/nokogiri_document.rb:
```diff
@@ -4,11 +4,12 @@
   include FieldSerializer
 
-  def initialize(noko)
+  def initialize(noko, content_class = nil)
```
NokogiriDocument should really be treated as if it were a separate standalone gem that _all_ scrapers can use, but which we just haven't extracted yet — i.e. you shouldn't be changing it unless that's going to become part of that gem. And it's definitely not clear to me from what's going on here what this extract functionality is or does or whether we'd need to include something like that there. What would the documentation for this be if it were moved up into a gem for every scraper?

tmtmtmtm commented on lib/term_page.rb:
```diff
@@ -3,12 +3,25 @@
   field :members do
-    MemberTable.new(table).members
+    MemberTable.new(constituency_table, constituency_member_class).members +
+    MemberTable.new(party_list_table, party_list_member_class).members
```
This seems like you should really be instantiating two different classes, e.g. ConstituencyMemberTable, and PartyListMemberTable.

ondenman committed:
Remove content_class property

A step towards breaking MemberTable into two separate classes: one
for constituency members, the other for party list members.

Since we are treating NokogiriDocument as a gem, we don't want to
change its functionality... the extra property is too specific to
this scraper to become part of the gem.

Instead of injecting the member class we want to create into
NokogiriDocument, we will create to separate table classes for each
type of member.

ondenman committed:
Add ConstituencyMemberTable

Class is responsible for scraping a list of constituency members.

ondenman committed:
Rename KhuralMember to ConstituencyMember

Since some members of the 2012 were elected from party lists, the
new class name -- ConstituencyMember -- makes the distinction
between constituency members and party list members.

ondenman committed:
fixup! Add ConstituencyMemberTable

ondenman committed:
Rename PartyListKhuralMember to PartyListMember

New name is more succint and consistent with class name: ConstituencyMember.

ondenman committed:
Add PartyListMemberTable

Class scrapes a list of party list members.

ondenman committed:
fixup! Add PartyListMemberTable

ondenman commented:
@tmtmtmtm 👀

tmtmtmtm commented:
@ondenman assuming this is still required, it's going to need rejigged in line with us now using `Scraped`…

ondenman commented:
This PR was superseded by https://github.com/everypolitician-scrapers/mongolia-khurai-wp-multiple-terms/pull/9
